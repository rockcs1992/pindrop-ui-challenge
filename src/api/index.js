import { API_URL } from '../utils/constants';

const fetchHandler = (url) =>
    fetch(url)
    .then((response) => response.json())
    .catch((error) => new Error(error));

const fetchMovies = (pageNum) => fetchHandler(`${API_URL}${pageNum++}`);

export { fetchMovies, fetchHandler };
