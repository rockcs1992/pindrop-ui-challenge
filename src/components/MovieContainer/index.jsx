import React from 'react';
import styled from 'styled-components';
import Poster from '../Poster';
import Loader from '../Loader';
import ErrorMessage from '../ErrorMessage';
import { IMG_BASE_URL } from '../../utils/constants';
import { fetchMovies } from '../../api';

const Grid = styled.div`
  background-color: black;
  display: grid;
  padding: 40px;
  grid-template-columns: repeat(4, 1fr);
  grid-template-rows: repeat(3, 1fr);
  grid-gap: 0;

  @media (max-width: 960px) {
    grid-template-columns: repeat(3, 1fr);
  }

  @media (max-width: 640px) {
    grid-template-columns: repeat(2, 1fr);
  }
`;

const Scroller = styled.div`
  width: 80%;
  height: 80%;
  overflow-y: scroll;
`;

class MovieContainer extends React.Component {
    constructor() {
        super();
        this.pageNum = 1;
        this.state = {
            movies: [],
            isLoading: false,
            error: false
        };
    }

    componentDidMount() {
        this.setState({ isLoading: true });
        this.getPopularMovies(this.pageNum++);
    }

    getPopularMovies = (page) => {
        fetchMovies(page)
        .then((data) => {
            const movies = data.results.map((result) => (
                {
                    posterPath: result.poster_path,
                    title: result.original_title,
                    id: result.id
                }
                ));
                this.setState((prevState) => ({
                    movies: [...prevState.movies, ...movies]
                }));
        })
        .catch(() => this.setState({ error: true }))
        .finally(() => this.setState({ isLoading: false }));
    }

    handleScroll = (e) => {
        const isBottom = e.target.scrollHeight - e.target.scrollTop === e.target.clientHeight;
        if (isBottom) {
            this.setState({
                isLoading: true,
                error: false
            });
            this.getPopularMovies(this.pageNum++);
        }
    }

    render() {
      const { movies, isLoading, error } = this.state;
      return (
        <Scroller onScroll={this.handleScroll}>
        {
            <React.Fragment>
                <Grid>
                {
                    movies.length && movies.map(({ posterPath, title, id }) =>
                        <Poster key={id} imgSrc={`${IMG_BASE_URL}${posterPath}`} title={title} />
                    )
                }
                </Grid>
                {isLoading && <Loader />}
                {error && <ErrorMessage />}
            </React.Fragment>
        }
        </Scroller>
      );
    }
}

export default MovieContainer;