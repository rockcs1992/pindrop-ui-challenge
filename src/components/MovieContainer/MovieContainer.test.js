import React from 'react';
import { render, mount } from 'enzyme';
import * as api from '../../api';
import MovieContainer from './index';

jest.mock('../../api');
api.fetchMovies.mockImplementation(() => Promise.resolve(true))

describe('onScroll', () => {
    it('should fire events when scrolled to the bottom', () => {
        const event = {
            target: {
                scrollHeight: 2,
                scrollTop: 1,
                clientHeight: 1
            }
        }
        const instance = mount(<MovieContainer />).instance();
        instance.pageNum = 1;
        instance.totalPages = 2;
        instance.handleScroll(event);
        expect(instance.state.isLoading).toBe(true);
        expect(instance.state.error).toBe(false);
        expect(api.fetchMovies).toHaveBeenCalledTimes(2);
    })

    it('should not fire events if not scrolled to the bottom', () => {
        const event = {
            target: {
                scrollHeight: 5,
                scrollTop: 1,
                clientHeight: 1
            }
        }
        const instance = mount(<MovieContainer />).instance();
        instance.pageNum = 1;
        instance.totalPages = 2;
        instance.handleScroll(event);
        expect(api.fetchMovies).not.toHaveBeenCalledTimes(2);
    })

    it('should not fire events if already on the last page', () => {
        const event = {
            target: {
                scrollHeight: 5,
                scrollTop: 1,
                clientHeight: 1
            }
        }
        const instance = mount(<MovieContainer />).instance();
        instance.pageNum = 3;
        instance.totalPages = 2;
        instance.handleScroll(event);
        expect(api.fetchMovies).not.toHaveBeenCalledTimes(2);
    })
})

describe('MovieContainer', () => {
    it('should render correctly', () => {
        const tree = render(<MovieContainer />);
        expect(tree).toMatchSnapshot();
    });

    it('should render correctly when loading', () => {
        const tree = mount(<MovieContainer />);
        tree.setState({ isLoading: true });
        expect(tree).toMatchSnapshot();
    });

    it('should render correctly when error', () => {
        const tree = mount(<MovieContainer />);
        tree.setState({ error: true });
        expect(tree).toMatchSnapshot();
    });
});

describe('On component mount', () => {
    it('should mount correctly', () => {
        const tree = mount(<MovieContainer />);
        expect(tree.instance().state.isLoading).toBe(true);
        expect(api.fetchMovies).toHaveBeenCalledWith(1);
    });
})

describe('handleMovieData', () => {
    const data = {
        results: [{
            poster_path: '/awd.jpg',
            original_title: 'test',
            id: '1200'
        }]
    };
    const expectedData = [{
        posterPath: '/awd.jpg',
        title: 'test',
        id: '1200'
    }]
    it('should handle movie data', () => {
        const instance = mount(<MovieContainer />).instance();
        instance.handleMovieData(data);
        expect(instance.state.isLoading).toBe(false);
        expect(instance.state.movies).toEqual(expectedData);
    })
})
