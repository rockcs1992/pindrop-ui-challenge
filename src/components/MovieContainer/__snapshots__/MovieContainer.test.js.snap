// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieContainer should render correctly 1`] = `
<div
  class="sc-htoDjs hGFiPY"
>
  <div
    class="sc-gzVnrw hhsMHj"
  >
    0
  </div>
</div>
`;

exports[`MovieContainer should render correctly when error 1`] = `
<MovieContainer>
  <styled.div
    onScroll={[Function]}
  >
    <div
      className="sc-htoDjs hGFiPY"
      onScroll={[Function]}
    >
      <styled.div>
        <div
          className="sc-gzVnrw hhsMHj"
        >
          0
        </div>
      </styled.div>
      <_default>
        <styled.div>
          <div
            className="sc-bxivhb EYjjH"
          >
            <styled.div>
              <div
                className="sc-ifAKCX idnxWS"
              />
            </styled.div>
          </div>
        </styled.div>
      </_default>
      <ErrorMessage>
        <styled.div>
          <div
            className="sc-bZQynM czlNrT"
          >
            <styled.h3>
              <h3
                className="sc-EHOje hmymRe"
              >
                Something went wrong
              </h3>
            </styled.h3>
          </div>
        </styled.div>
      </ErrorMessage>
    </div>
  </styled.div>
</MovieContainer>
`;

exports[`MovieContainer should render correctly when loading 1`] = `
<MovieContainer>
  <styled.div
    onScroll={[Function]}
  >
    <div
      className="sc-htoDjs hGFiPY"
      onScroll={[Function]}
    >
      <styled.div>
        <div
          className="sc-gzVnrw hhsMHj"
        >
          0
        </div>
      </styled.div>
      <_default>
        <styled.div>
          <div
            className="sc-bxivhb EYjjH"
          >
            <styled.div>
              <div
                className="sc-ifAKCX idnxWS"
              />
            </styled.div>
          </div>
        </styled.div>
      </_default>
    </div>
  </styled.div>
</MovieContainer>
`;
